Correct the order of module configurations

MikeAlhayek opened the pull request:
Fix #14874

In the PR, we have introduced a central constants class that manages the order of module registrations. This improvement provides a comprehensive view of all registered modules, making it easier to adjust the order as needed. If someone needs to change the order of any module, they can see at a glance if other order numbers need to be adjusted accordingly.

MikeAlhayek commented on src/OrchardCore/OrchardCore.Abstractions/OrchardCoreConstants.cs:
```diff
@@ -28,4 +28,44 @@ public class Roles
+        // The order number should always be greater than data protection modules. 
+        public const int AzureDataProtection = DataProtection + 10;
+
+        public const int HomeRoute = -150;
```
Self-note: the home route probably should be greater than the Authentication. Did not change it here to avoid possible unknown issues. so maybe -10 is sufficient. Not really sure why this module even need to define an order and it does not override services. Probably `Default` 0 is even better

kevinchalet commented on src/OrchardCore/OrchardCore.Abstractions/OrchardCoreConstants.cs:
```diff
@@ -28,4 +28,44 @@ public class Roles
+        // The reverse proxy should always be configured before the 'Authentication' and security initialization logic.
+        public const int ReverseProxy = Authentication * 2;
+
+        public const int Cors = ReverseProxy;
```
CORS should be registered after the reverse proxy module to ensure the correct host is used.

MikeAlhayek replied:
@kevinchalet I made this change. Any other suggestion or you good with this?

kevinchalet reviewed (approved):
👍🏻

Piedone reviewed (changes requested):
I've swapped out the preview packages under [Lombiq's Open-Source Orchard Core Extensions, this PR](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/pull/780) and has run its tests. Those passed, so this is a good sign.

MikeAlhayek commented on src/OrchardCore/OrchardCore.Abstractions/OrchardCoreConstants.cs:
same here. But I think it is important for these key order to be in the same place so we can easily identify any issue when we make order changes.

Piedone replied:
Yeah, let's keep it like this.

MikeAlhayek reviewed (commented):
cleanup

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs:
```diff
@@ -20,7 +20,8 @@ public Startup(IShellConfiguration configuration, ILogger<Startup> logger)
     }
 
     // Assume that this module will override default configuration, so set the Order to a value above the default.
```
indeed. Removed.

kevinchalet commented:
I don't see `OrchardCore.Security` in the list of changed files: it doesn't have an explicit order, but we should probably give it one while we're at it. Probably just before or after CORS (but after the reverse proxy module)

MikeAlhayek commented:
> I don't see OrchardCore.Security in the list of changed files: it doesn't have an explicit order, but we should probably give it one while we're at it. Probably just before or after CORS (but after the reverse proxy module)

@kevinchalet not a bad idea. I added that at the same order as CORS just to show that there is no significant difference in order between CORS and Reverse Proxy. I added a note as it must configured after the reverse proxy.

kevinchalet commented:
> @kevinchalet not a bad idea. I added that at the same order as CORS just to show that there is no significant difference in order between CORS and Reverse Proxy. I added a note as it must configured after the reverse proxy.

👍🏻 

That said, I wonder if we shouldn't use unique values to make the order more explicit: I haven't tested, but since the ordering logic is just an `.OrderBy(x => x.ConfigureOrder)` on the list of `IStartup` services, I'd assume the actual order of modules having the same value actually depends on their position in the `IServiceCollection` instance (when using with LINQ to Objects, `.OrderBy(...)` performs a stable sort, but it depends on the ordering logic of the input collection).

MikeAlhayek commented:
IMO, if there is a need to place one after another, then we should use unique values to guaranty the order. But, if there we don't care if CORS came before Security or the other way around, then I don't see the value of adding a unique value. In other words, if either Security or CORS were registered first, there should not be any issue.
